regcomp_internal.h: new -Wtypedef-redefinition build-time warnings

A recent commit has introduced 5 instances of `-Wtypedef-redefinition` build-time warnings, as observed when doing a threaded build on FreeBSD-12 using `clang` (default version 10.0.1) as the C-compiler.

The 5 instances each appear like this in my `make`-output parsing program:
```
  {
    char   => 3,
    group  => "Wtypedef-redefinition",
    line   => 626,
    source => "./regcomp_internal.h",
    text   => "redefinition of typedef 'scan_data_t' is a C11 feature",
  },
```
In context, the first of the 5 instances:
```
In file included from regcomp.c:146:
./regcomp_internal.h:626:3: warning: redefinition of typedef 'scan_data_t' is a C11 feature [-Wtypedef-redefinition]
} scan_data_t;
  ^
./perl.h:4525:28: note: previous definition is here
typedef struct scan_data_t scan_data_t;
                           ^
1 warning generated.
```
The 5 instances differ only with respect to their respective `file included from` locations:
```
:!zgrep -n -B1 'warning:' 4a11c5038c.freebsd.threaded.clang.maketp.output.txt.gz
12-In file included from regcomp.c:146:
13:./regcomp_internal.h:626:3: warning: redefinition of typedef 'scan_data_t' is a C11 feature [-Wtypedef-redefinition]
--
30-In file included from regcomp_invlist.c:19:
31:./regcomp_internal.h:626:3: warning: redefinition of typedef 'scan_data_t' is a C11 feature [-Wtypedef-redefinition]
--
39-In file included from regcomp_study.c:19:
40:./regcomp_internal.h:626:3: warning: redefinition of typedef 'scan_data_t' is a C11 feature [-Wtypedef-redefinition]
--
48-In file included from regcomp_trie.c:19:
49:./regcomp_internal.h:626:3: warning: redefinition of typedef 'scan_data_t' is a C11 feature [-Wtypedef-redefinition]
```
Since most of the above files are very new to the core distribution, these warnings must have been first triggered in the following commit or later:
```
commit 85900e28cc250e1c4603f11073b77d0c6b5cff46
Author:     Yves Orton <demerphq@gmail.com>
AuthorDate: Fri Dec 9 10:00:17 2022
Commit:     Yves Orton <demerphq@gmail.com>
CommitDate: Fri Dec 9 15:19:29 2022

    regcomp.c - decompose into smaller files
    
    This splits a bunch of the subcomponents of the regex engine into
    smaller files.
    
           regcomp_debug.c
           regcomp_internal.h
           regcomp_invlist.c
           regcomp_study.c
           regcomp_trie.c
```
@demerphq, can you take a look?  Thanks.

Sure. but alas, I wont have much time for it over the next few days.

Yves

On Wed, 14 Dec 2022 at 22:54, James E Keenan ***@***.***>
wrote:

> A recent commit has introduced 5 instances of -Wtypedef-redefinition
> build-time warnings, as observed when doing a threaded build on FreeBSD-12
> using clang (default version 10.0.1) as the C-compiler.
>
> The 5 instances each appear like this in my make-output parsing program:
>
>   {
>     char   => 3,
>     group  => "Wtypedef-redefinition",
>     line   => 626,
>     source => "./regcomp_internal.h",
>     text   => "redefinition of typedef 'scan_data_t' is a C11 feature",
>   },
>
> In context, the first of the 5 instances:
>
> In file included from regcomp.c:146:
> ./regcomp_internal.h:626:3: warning: redefinition of typedef 'scan_data_t' is a C11 feature [-Wtypedef-redefinition]
> } scan_data_t;
>   ^
> ./perl.h:4525:28: note: previous definition is here
> typedef struct scan_data_t scan_data_t;
>                            ^
> 1 warning generated.
>
> The 5 instances differ only with respect to their respective file
> included from locations:
>
> :!zgrep -n -B1 'warning:' 4a11c5038c.freebsd.threaded.clang.maketp.output.txt.gz
> 12-In file included from regcomp.c:146:
> 13:./regcomp_internal.h:626:3: warning: redefinition of typedef 'scan_data_t' is a C11 feature [-Wtypedef-redefinition]
> --
> 30-In file included from regcomp_invlist.c:19:
> 31:./regcomp_internal.h:626:3: warning: redefinition of typedef 'scan_data_t' is a C11 feature [-Wtypedef-redefinition]
> --
> 39-In file included from regcomp_study.c:19:
> 40:./regcomp_internal.h:626:3: warning: redefinition of typedef 'scan_data_t' is a C11 feature [-Wtypedef-redefinition]
> --
> 48-In file included from regcomp_trie.c:19:
> 49:./regcomp_internal.h:626:3: warning: redefinition of typedef 'scan_data_t' is a C11 feature [-Wtypedef-redefinition]
>
> Since most of the above files are very new to the core distribution, these
> warnings must have been first triggered in the following commit or later:
>
> commit 85900e28cc250e1c4603f11073b77d0c6b5cff46
> Author:     Yves Orton ***@***.***>
> AuthorDate: Fri Dec 9 10:00:17 2022
> Commit:     Yves Orton ***@***.***>
> CommitDate: Fri Dec 9 15:19:29 2022
>
>     regcomp.c - decompose into smaller files
>
>     This splits a bunch of the subcomponents of the regex engine into
>     smaller files.
>
>            regcomp_debug.c
>            regcomp_internal.h
>            regcomp_invlist.c
>            regcomp_study.c
>            regcomp_trie.c
>
> @demerphq <https://github.com/demerphq>, can you take a look? Thanks.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Perl/perl5/issues/20608>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAZ5R2JTV7CPGA5KSKYVJDWNI6ZFANCNFSM6AAAAAAS67K66U>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

--- a/regcomp_internal.h
+++ b/regcomp_internal.h
@@ -599,7 +599,8 @@ struct scan_data_substrs {
     I32 flags;          /* per substring SF_* and SCF_* flags */
 };
 
-typedef struct scan_data_t {
+/* this is typedef'ed in perl.h */
+struct scan_data_t {
     /*I32 len_min;      unused */
     /*I32 len_delta;    unused */
     SSize_t pos_min;
@@ -623,7 +624,7 @@ typedef struct scan_data_t {
                                  pointer - the op may have been optimized
                                  away */
     regnode_ssc *start_class;
-} scan_data_t;
+};
 
 /*
  * Forward declarations for pregcomp()'s friends.

